framework/src/base/MooseApp.C

```diff
@@ -438,3 +452,205 @@ MooseApp::getOutputWarehouse()
+
+  char * moose_lib_path_env = std::getenv("MOOSE_LIBRARY_PATH");
+  if (moose_lib_path_env)
+  {
+    std::string moose_lib_path(moose_lib_path_env);
+    std::vector<std::string> tmp_paths;
+
+    MooseUtils::tokenize(moose_lib_path, tmp_paths, 1, ":");
+
+    // merge the two vectors together (all possible search paths)
+    paths.insert(paths.end(), tmp_paths.begin(), tmp_paths.end());
+  }
+
+  // Attempt to dynamically load the library
+  for (std::vector<std::string>::const_iterator path_it = paths.begin(); path_it != paths.end(); ++path_it)
+    if (MooseUtils::checkFileReadable(*path_it + '/' + library_name, false, false))
+      loadLibraryAndDependencies(*path_it + '/' + library_name, registration_method, factory);
+    else
+      mooseWarning("Unable to open library file \"" << *path_it + '/' + library_name << "\". Double check for spelling errors.");
+}
+
+void
+MooseApp::loadLibraryAndDependencies(const std::string & library_filename, const std::string & registration_method_name, Factory * factory)
+{
+  std::string line;
+  std::string dl_lib_filename;
+
+  // This RE looks for absolute path libtool filenames (i.e. begins with a slash and ends with a .la)
+  pcrecpp::RE re_deps("(/\\S*\\.la)");
+
+  std::ifstream handle(library_filename.c_str());
+  if (handle.is_open())
+  {
+    while (std::getline(handle, line))
+    {
+      // Look for the system dependent dynamic library filename to open
+      if (line.find("dlname=") != std::string::npos)
+        // Magic numbers are computed from length of this string "dlname=' and line minus that string plus quotes"
+        dl_lib_filename = line.substr(8, line.size()-9);
+
+      if (line.find("dependency_libs=") != std::string::npos)
+      {
+        pcrecpp::StringPiece input(line);
+        pcrecpp::StringPiece depend_library;
+        while (re_deps.FindAndConsume(&input, &depend_library))
+          // Recurse here to load dependent libraries in depth-first order
+          loadLibraryAndDependencies(depend_library.as_string(), registration_method_name, factory);
+
+        // There's only one line in the .la file containing the dependency libs so break after finding it
+        break;
+      }
+    }
+    handle.close();
+  }
+
+  // Time to load the library, First see if we've already loaded this particular dynamic library
+  if (_lib_handles.find(library_filename) == _lib_handles.end() && // make sure we haven't already loaded this library
+      dl_lib_filename != "")                                       // AND make sure we have a library name (we won't for static linkage)
+  {
+    std::pair<std::string, std::string> lib_name_parts = MooseUtils::splitFileName(library_filename);
+
+    // Assemble the actual filename using the base path of the *.la file and the dl_lib_filename
+    std::string dl_lib_full_path = lib_name_parts.first + '/' + dl_lib_filename;
+
+    void * handle = dlopen(dl_lib_full_path.c_str(), RTLD_LAZY);
```

Comment:
If this is the first `dlopen` dependency we have in MOOSE, we need to be sure it fails gracefully (instead of fails to compile) on systems that don't support dlopen.  And yes, there are some out there, @dschwen ran into one with his JIT stuff.